### ISSUE TITLE
In relation to unknowncategory #136

### ISSUE THREAD

**Issue by rimacher** [CLAMP-IT/moodle-blocks_filtered_course_list#139]
Hello and thank you for the wonderful plugin,

I was happy to see issue unknowncategory #136 fixed but there is a problem with the category filters' permissions.

It appears that teachers can't see their courses if the category is hidden while they should be able to and only the students shouldn't. Students also should be able to see visible courses in hidden categories.

**Comment by kwiliarty** [comment]
Hello @rimacher .

To test what you are describing I have created a hidden category with a visible child category. In the hidden category I have created a hidden course and a visible course. In the visible child of the hidden category I have also created a visible course and a hidden course. Schematically:

```
Hidden Category
  - Visible Course in Hidden Category
  - Hidden Course in Hidden Category
  Visible Child of Hidden Category
    - Visible Course in Visible Child of Hidden Category
    - Hidden Course in Visible Child of Hidden Category
```

In each course I enrolled a teacher and a student. Below are the screenshots for each user showing the Filtered Course List and the Moodle-native Courses block for comparison. As far as I can tell, the Filtered Course List is exhibiting the behavior your are looking for (and the same behavior as the native block).

First the teacher view:

<img width="188" alt="teacher-view" src="https://user-images.githubusercontent.com/407636/58805077-50b9cb00-85e1-11e9-9f07-0da2c0f3a919.png">

Then the student view:

<img width="187" alt="student-view" src="https://user-images.githubusercontent.com/407636/58805390-fcfbb180-85e1-11e9-976e-f182d72c7dff.png">

Are you seeing something different? Which Moodle and block versions are you using? I am using Moodle 3.7+ and Filtered Course List 4.1.3.

**Comment by rimacher** [comment]
I'm using Moodle 3.4+ and Filtered Course List v3.5.0.

I did the following steps:
1. My category tree is:  **A**>**B**>**C**>**D** and all of them are visible.
![04-06-2019 12-41-41](https://user-images.githubusercontent.com/16016842/58869059-2c83ea00-86c6-11e9-9c4e-1555b13aa85f.png)

2. Filter setting: **category | exp | A | 0**

3. A teacher is enrolled in two courses in category **D**. One course is visible the other is hidden. All works OK.
![04-06-2019 13-04-23](https://user-images.githubusercontent.com/16016842/58870566-1af01180-86c9-11e9-862b-e995dbe23216.png)


4. Change the field "visible" of category A to 0.
![04-06-2019 12-57-56](https://user-images.githubusercontent.com/16016842/58870162-5211f300-86c8-11e9-8c1e-d218c3d5953c.png)

5. Teacher no longer sees courses.
![04-06-2019 13-03-38](https://user-images.githubusercontent.com/16016842/58870582-22171f80-86c9-11e9-81fe-069bcc2fd021.png)

6. Change the filler setting to: **category | exp | B | 0** and courses appear again.

**Comment by kwiliarty** [comment]
Thanks for your patience with this. I can reproduce the problem now. I must have been doing something wrong in my own test, because now I cannot reproduce my own result. 🤦‍♂ I'll try to figure out what's going on. One thing I notice is that in my automated test I am setting the category filter id to the visible child of the hidden parent. I'll need a new test that covers the case when the hidden category is the main node or a descendant of the node referenced in the block configuration.

**Comment by azito122** [comment]
My interest was piqued by this issue, so I tried reproducing it locally and saw that the courses missing from the display were in fact visible under the Other courses tab. If my understanding of what's happening here is correct, I think this is the same thing as seen in issue #102

**Comment by kwiliarty** [comment]
Yeah, it has to do with the recursive function we use to walk the category tree. It calls `core_course_category::get($catid)->get_children()`, but that function only returns child categories visible to the current user, so when we hit a hidden child, we stop following that branch. I'm contemplating alternatives.

As you note, we're finding all the courses. But since we don't traverse the whole category tree courses under a hidden node are winding up in "Other courses" (unless they get picked up by some other filter). At any rate, they're not appearing where they should.

You can, incidentally, target a visible category below a hidden category in the filter configuration. That works as of #136. As long as there is no hidden category between the starting point and the endpoint. That's @rimacher 's `category | exp | B | 0` example.

**Comment by kwiliarty** [comment]
Just trying to spell out the target behavior here. Let's imagine this category structure:

```
Miscellaneous
  --Hidden
    --Visible
```

Further imagine that there are courses in each category. There are three scenarios: 

1. We target a visible category within a hidden category: The block should create a rubric for the Visible category and display its courses to the user.
2. We target the Hidden category: The block should create rubrics for any visible children within the specified depth. The block should not create a rubric for the hidden category, but its courses should appear under "Other courses" if they are accessible to the user.
3. We target a category above the Hidden category. The block should create rubrics for all visible children within the specified depth including visible children of hidden categories. It should omit rubrics for hidden categories, but any courses accessible to the user should appear under "Other courses"

In addition: Hidden categories should not show up when references via `PARENT` or `ANCESTRY` tokens in the category rubric template.

**Comment by kwiliarty** [comment]
@rimacher I have resolved the issue for the master branch, and I am in the process of backporting it for Moodle 3.4 and 3.5. I'll release fixed branches in a day or two.

**Comment by kwiliarty** [comment]
@rimacher This is resolved for Moodle 3.6 and 3.7 in version v4.1.4. For Moodle 3.4 and 3.5 it is resolved in v3.5.1. Both fix versions have been added to the Moodle plugins database.
